### PR TITLE
Allow ARM on macOS

### PR DESCRIPTION
**Pull request by fxcoudert**
Although we're not entirely there yet, the time for Apple Silicon support is nearing. I've had a look at the installer script to see what is needed there.

### CODE REVIEW

**Comment by fxcoudert** [comment]
Around line 425, there is:

```
directories=(bin etc include lib sbin share var opt
             share/zsh share/zsh/site-functions
             var/homebrew var/homebrew/linked
             Cellar Caskroom Homebrew Frameworks)
mkdirs=()
for dir in "${directories[@]}"; do
  if ! [[ -d "${HOMEBREW_PREFIX}/${dir}" ]]; then
    mkdirs+=("${HOMEBREW_PREFIX}/${dir}")
  fi
done
```

I think the `Homebrew` directory here is an error. `HOMEBREW_REPOSITORY` should be created with a different logic.

**Comment by fxcoudert** [comment]
Running this version on Apple Silicon currently fails because we're not giving `/opt/homebrew` the right ownership. It gets `root:wheel`, which is not appropriate. Whatever else happens, we always need to be owners of `HOMEBREW_REPOSITORY`.

```
brewadmin@administrators-Mac-5 ~ % bash ./install.sh    
==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew
==> The following new directories will be created:
/opt/homebrew/bin
/opt/homebrew/etc
/opt/homebrew/include
/opt/homebrew/lib
/opt/homebrew/sbin
/opt/homebrew/share
/opt/homebrew/var
/opt/homebrew/opt
/opt/homebrew/share/zsh
/opt/homebrew/share/zsh/site-functions
/opt/homebrew/var/homebrew
/opt/homebrew/var/homebrew/linked
/opt/homebrew/Cellar
/opt/homebrew/Caskroom
/opt/homebrew/Homebrew
/opt/homebrew/Frameworks

Press RETURN to continue or any other key to abort
==> /usr/bin/sudo /bin/mkdir -p /opt/homebrew
==> /usr/bin/sudo /usr/sbin/chown root:wheel /opt/homebrew
==> /usr/bin/sudo /bin/mkdir -p /opt/homebrew/bin /opt/homebrew/etc /opt/homebrew/include /opt/homebrew/lib /opt/homebrew/sbin /opt/homebrew/share /opt/homebrew/var /opt/homebrew/opt /opt/homebrew/share/zsh /opt/homebrew/share/zsh/site-functions /opt/homebrew/var/homebrew /opt/homebrew/var/homebrew/linked /opt/homebrew/Cellar /opt/homebrew/Caskroom /opt/homebrew/Homebrew /opt/homebrew/Frameworks
==> /usr/bin/sudo /bin/chmod g+rwx /opt/homebrew/bin /opt/homebrew/etc /opt/homebrew/include /opt/homebrew/lib /opt/homebrew/sbin /opt/homebrew/share /opt/homebrew/var /opt/homebrew/opt /opt/homebrew/share/zsh /opt/homebrew/share/zsh/site-functions /opt/homebrew/var/homebrew /opt/homebrew/var/homebrew/linked /opt/homebrew/Cellar /opt/homebrew/Caskroom /opt/homebrew/Homebrew /opt/homebrew/Frameworks
==> /usr/bin/sudo /usr/sbin/chown brewadmin /opt/homebrew/bin /opt/homebrew/etc /opt/homebrew/include /opt/homebrew/lib /opt/homebrew/sbin /opt/homebrew/share /opt/homebrew/var /opt/homebrew/opt /opt/homebrew/share/zsh /opt/homebrew/share/zsh/site-functions /opt/homebrew/var/homebrew /opt/homebrew/var/homebrew/linked /opt/homebrew/Cellar /opt/homebrew/Caskroom /opt/homebrew/Homebrew /opt/homebrew/Frameworks
==> /usr/bin/sudo /usr/bin/chgrp admin /opt/homebrew/bin /opt/homebrew/etc /opt/homebrew/include /opt/homebrew/lib /opt/homebrew/sbin /opt/homebrew/share /opt/homebrew/var /opt/homebrew/opt /opt/homebrew/share/zsh /opt/homebrew/share/zsh/site-functions /opt/homebrew/var/homebrew /opt/homebrew/var/homebrew/linked /opt/homebrew/Cellar /opt/homebrew/Caskroom /opt/homebrew/Homebrew /opt/homebrew/Frameworks
==> Downloading and installing Homebrew...
/opt/homebrew/.git: Permission denied
Failed during: git init -q
```

File: install.sh
Comment (MikeMcQuaid) [inline install.sh:39]:
```suggestion
  # and ~/.linuxbrew (which is unsupported) if run interactively.
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far!

**Comment by MikeMcQuaid** [comment]
> I think the `Homebrew` directory here is an error. `HOMEBREW_REPOSITORY` should be created with a different logic.

Yes, that could reference `HOMEBREW_REPOSITORY` if needed or just be removed perhaps.

**Comment by fxcoudert** [comment]
I've tested on ARM and Intel macOS. I'd like to have some Linux testing, but otherwise I think this is complete and ready for review.

**Comment by sjackman** [comment]
> I'd like to have some Linux testing, but otherwise I think this is complete and ready for review.

I gave it a quick test on Linux in `/home/linuxbrew/.linuxbrew` with and without `sudo` available, and it worked for me.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good to me! Let's hold off merging until we're able to announce we "support" this (i.e. we have a non-trivial number of bottles).

**Comment by billinghamj** [comment]
I think this is great, as really a shockingly large amount of stuff "just works" now - it's really impressive how fast that has happened.

But I wonder if you might want to consider describing it as more partial support, or support with significant caveats? As some not-insignificant proportion of formulae will likely not be compatible for a _very_ long time, if ever. So setting user expectations in some form may help?

**Comment by MikeMcQuaid** [comment]
> But I wonder if you might want to consider describing it as more partial support, or support with significant caveats? As some not-insignificant proportion of formulae will likely not be compatible for a long time, if ever. So setting user expectations in some form may help?

Yup, that's the plan before we'd merge this 👍🏻

**Comment by fxcoudert** [comment]
We have passed 1000 ARM bottles, and I would like to merge this and enable ARM testing on CI: https://github.com/Homebrew/homebrew-core/pull/67577

**Review by MikeMcQuaid — COMMENTED** [review]
Can we print some sort of disclaimer that there's still a lot of packages missing bottles and recommend Rosetta 2 for now? What's the state of `brew doctor` on a default ARM setup; might want to address/add to that too?

**Comment by fxcoudert** [comment]
Re. `brew doctor`: https://github.com/Homebrew/brew/pull/10132

**Comment by MikeMcQuaid** [comment]
> Re. `brew doctor`: [Homebrew/brew#10132](https://github.com/Homebrew/brew/pull/10132)

👍🏻 to merge once that's merged.

**Comment by fxcoudert** [comment]
> 👍🏻 to merge once that's merged

@MikeMcQuaid I believe we need a version shipped, otherwise people will get brew doctor complaints

**Comment by fxcoudert** [comment]
I've opened a brew issue tracking Apple Silicon support, and stating our current level of support, here: https://github.com/Homebrew/brew/issues/10152

**Comment by richiksc** [comment]
There's a new `brew` version released now with Homebrew/brew#10132 included. Can this be merged now?

**Comment by fievelk** [comment]
Just a quick note: since this PR got merged, I think it would be good to updated the first lines of the instructions on https://docs.brew.sh/Installation (they only mention `/usr/local` for the script installation).

**Comment by fxcoudert** [comment]
@fievelk indeed, thank you for noticing! https://github.com/Homebrew/brew/pull/10182

**Comment by charliegroll** [comment]
@fievelk @fxcoudert it also says that a 64-bit Intel CPU is required 😉

**Comment by fxcoudert** [comment]
@charliegroll my pull request also changes that part: https://github.com/Homebrew/brew/pull/10182

**Comment by nickoe** [comment]
@fxcoudert Why does the `HOMEBREW_PREFIX` need to be different depending on arch?

**Comment by fxcoudert** [comment]
@nickoe see https://github.com/Homebrew/brew/issues/9177

**Comment by nickoe** [comment]
@fxcoudert That issue does not really talk about why. I guess the closest answer is https://github.com/Homebrew/brew/issues/9130#issuecomment-731997309

But as lenlo mentions:
> [...]
> I am less happy with the heavy handed way /opt/homebrew suddenly got usurped for ARM binaries only -- even on x86 systems that will never have any ARM binaries on them. 
> [...]

But yet, that does not explain why they prefixes has to be different.

**Comment by fxcoudert** [comment]
> In the initial phase of Apple Silicon migration, in order to avoid using the same prefix for both Intel and ARM binaries, we have two prefixes instead: /usr/local for Intel, and /opt/homebrew for Apple Silicon

Basically because not everything supports Apple Silicon, we need to have coexisting installations for Rosetta and native. Hence we cannot use `/usr/local` for ARM

**Comment by nickoe** [comment]
@fxcoudert but couldn't hat just be a subdir in the prefix?

**Comment by sjackman** [comment]
A subdirectory like `/usr/local/arm` was one of proposals considered, but `/opt/homebrew` was preferred by the majority.

**Comment by richiksc** [comment]
Off-topic, but I prefer `/opt/homebrew` as well, because it creates separation between binaries installed by Homebrew, and binaries installed manually to `/usr/local/bin`. If I ever wanted to delete all Homebrew installed binaries or in fact Homebrew itself, I can easily run `rm -rf /opt/homebrew[/bin]` without worrying about anything else installed to `/usr/local`.

**Comment by sjackman** [comment]
Homebrew on Linux uses `/home/linuxbrew/.linuxbrew` as the default prefix for this reason among others.

**Pull request merged**
